src/swarm/neo/client/mixins/ClientCore.d

```diff
@@ -400,6 +402,114 @@ template ClientCore ( )
+                getter
+            Getter = type of stats getter struct (e.g. the Stats class defined
+                above or a derived class)
+            getter = instance of stats getter Getter
+            logger = stats log to write the filled instance of Aggr to
+
+    ***************************************************************************/
+
+    private void logStatsFromAggregate ( Aggr, Getter )
+        ( Getter getter, StatsLog logger )
+    {
+        static assert(is(Aggr == struct));
+
+        Aggr aggr;
+
+        foreach ( i, ref field; aggr.tupleof )
+            mixin("field = getter." ~ FieldName!(i, Aggr) ~ "();");
+        logger.add(aggr);
+    }
+
+    ///
+    unittest
+    {
+        void logClientStats ( typeof(this) client, StatsLog logger )
+        {
+            struct StatsAggregate
+            {
+                size_t num_active_requests;
+                size_t max_active_requests;
+                double active_requests_fraction;
+                size_t num_registered_nodes;
+                size_t num_initializing_nodes;
+                double initializing_nodes_fraction;
+                size_t num_connected_nodes;
+                bool all_nodes_connected;
+                double connected_nodes_fraction;
+            }
+
+            scope stats = client.new Stats;
+
+            // In D2, we can ensure that the aggregate struct isn't missing any
+            // of the stats values returned by the getter.
+            version ( D_Version2 )
+                checkStatsAggregate!(StatsAggregate)(stats);
+
+            client.logStatsFromAggregate!(StatsAggregate)(stats, logger);
+        }
+    }
+
+    /***************************************************************************
+
+        D2-only function to check that the stats aggregate struct contains a
+        field of the same type as each primitive-returning method in the stats
+        getter. May be called by a client to ensure that all stats provided by
+        the getter are written to the stats log by logStatsFromAggregate().
+
+        Params:
+            Aggr = stats aggregate struct with one field per stats to log. The
+                names and types of the fields are expected to match the names
+                and reurn types of the getter methods of the provided stats
+                getter
+            Getter = type of stats getter struct (e.g. the Stats class defined
+                above or a derived class)
+            getter = instance of stats getter Getter
+
+    ***************************************************************************/
+
+    version ( D_Version2 )
+    private static void checkStatsAggregate ( Aggr, Getter ) ( Getter getter )
+    {
+        mixin(`
+            enum object_members = __traits(allMembers, Object);
+            foreach ( member; __traits(allMembers, Getter) )
+            {
+                static if ( !__traits(hasMember, Object, member)
+                    && member != "this" )
```

Comment:
I wouldn't remove this test from here, tbh. It's just a generalization of the `version(some_sanity_check_flag)` sanity check helper, nothing depends on it, it can be simply compiled out, and it's just an extra check not affecting both D1 nor D2 version.